File: packages/gbnf-experiment/data/20260909T002206Z_273cf6b9/ported_implementation/src/grammar-graph/grammar-graph-types.ts
import type { GraphPointer } from './graph-pointer.ts';
import type { Pointers } from './pointers.ts';
import { RuleRef } from './rule-ref.ts';

export type Colorize = (text: string | number, color: string) => string;

export interface PrintOpts {
  pointers?: Pointers;
  colorize: Colorize;
  show_position?: boolean;
}

export type Range = [number, number];

export type RuleDict = { type: string; value?: unknown };

export class Rule {
  readonly type: string = 'Rule';

  equals(other: unknown): boolean {
    return other instanceof (this.constructor as new (...args: never[]) => unknown);
  }

  toString(): string {
    return `${this.type}()`;
  }

  /** Mirrors the reference implementation's `__dict__`, used for serializing a rule. */
  get dict(): RuleDict {
    return {
      type: this.type,
    };
  }
}

export class RuleWithValue extends Rule {
  override readonly type: string = 'RuleWithValue';
  value: unknown;

  constructor(value: unknown) {
    super();
    this.value = value;
  }

  override equals(other: unknown): boolean {
    return super.equals(other) && valuesAreEqual(this.value, (other as RuleWithValue).value);
  }

  override toString(): string {
    return `${this.type}(value=${JSON.stringify(this.value)})`;
  }

  override get dict(): RuleDict {
    return {
      ...super.dict,
      value: this.value,
    };
  }
}

export class RuleWithListOfIntsOrRanges extends RuleWithValue {
  override readonly type: string = 'RuleWithListOfIntsOrRanges';
  declare value: (number | Range)[];

  constructor(value: (number | Range)[] = []) {
    super([...value]);
  }
}

export class RuleChar extends RuleWithListOfIntsOrRanges {
  override readonly type = 'RuleChar';
}

export class RuleCharExclude extends RuleWithListOfIntsOrRanges {
  override readonly type = 'RuleCharExclude';
}

export class RuleEnd extends Rule {
  override readonly type = 'RuleEnd';
}

export type UnresolvedRule = RuleChar | RuleCharExclude | RuleRef | RuleEnd;

// ValidInput can either be a string, or a number indicating a code point.
// It CANNOT be a number representing a number; a number intended as input (like "8")
// should be passed in as a string.
export type ValidInput = string | number | number[];

// RuleRefs should never be exposed to the end user.
export type ResolvedRule = RuleCharExclude | RuleChar | RuleEnd;

export type ResolvedGraphPointer = GraphPointer<ResolvedRule>;

const valuesAreEqual = (left: unknown, right: unknown): boolean => {
  if (Array.isArray(left) && Array.isArray(right)) {
    return left.length === right.length && left.every((value, idx) => valuesAreEqual(value, right[idx]));
  }
  return left === right;
};

export { RuleRef };

File: proposed-projects/round-trip-experiment/reverse/20260909T175143Z_62b61540/20260910T094302Z_8f8be025/ported_implementation/gbnf/grammar_graph/grammar_graph_types.py
from enum import Enum
from typing import Union

# A range of code points, as a two element [start, end] list.
Range = list[int]

# ValidInput can either be a string, or a number indicating a code point.
# It CANNOT be a number representing a number; a number intended as input (like "8")
# should be passed in as a string.
ValidInput = Union[str, int, list[int]]


class RuleType(str, Enum):
    CHAR = "char"
    CHAR_EXCLUDE = "char_exclude"
    END = "end"
    REF = "ref"


class RuleChar:
    def __init__(self, value: list[Union[int, Range]]):
        self.type = RuleType.CHAR
        self.value: list[Union[int, Range]] = [*value]

    def __eq__(self, other: object) -> bool:
        if type(other) is not RuleChar:
            return NotImplemented
        return self.value == other.value

    # Rules are held in identity keyed collections while walking the graph, so
    # hashing stays identity based even though equality is by value.
    __hash__ = object.__hash__

    def __repr__(self) -> str:
        return f"RuleChar({self.value!r})"


class RuleCharExclude:
    def __init__(self, value: list[Union[int, Range]]):
        self.type = RuleType.CHAR_EXCLUDE
        self.value: list[Union[int, Range]] = [*value]

    def __eq__(self, other: object) -> bool:
        if type(other) is not RuleCharExclude:
            return NotImplemented
        return self.value == other.value

    __hash__ = object.__hash__

    def __repr__(self) -> str:
        return f"RuleCharExclude({self.value!r})"


class RuleEnd:
    def __init__(self) -> None:
        self.type = RuleType.END

    def __eq__(self, other: object) -> bool:
        if type(other) is not RuleEnd:
            return NotImplemented
        return True

    __hash__ = object.__hash__

    def __repr__(self) -> str:
        return "RuleEnd()"


class PrintOpts:
    def __init__(self, colorize, pointers=None, show_position: bool = False):
        self.colorize = colorize
        self.pointers = pointers
        self.show_position = show_position
